### PR TITLE
feat(incremental-planner): add FetchGraph with undo-log rollback

### PR DESCRIPTION
**Pull request by tninesling**
## Summary

Adds `FetchGraph`, the mutable graph of fetch groups (subgraph calls) built incrementally during BULB search.

The key design choice is an **undo log** instead of clone-per-branch: `checkpoint()` records the log position, mutations append ops, and `rollback(cp)` reverses them in LIFO order. This keeps trial branches cheap (no graph cloning during search) while supporting arbitrary rollback depth for BULB backtracking.

Each node is a `FetchNode` (subgraph name, group kind, selection builder); edges carry `InputContribution` (entity key conditions and rewrite info). Pipeline depth and per-stage counts are maintained incrementally so `cost()` is O(max_depth) instead of a full graph walk.

Constants `FETCH_COST` and `PIPELINING_COST` are defined at module level and shared by `cost()`, the plan builder (next PR), and the test oracle.

## Test plan

- Unit tests verify: empty graph cost, checkpoint/rollback state preservation, nested checkpoints, incremental cost matches from-scratch recomputation through mixed add/raise/rollback sequences, clone independence, merge-at-root

### CODE REVIEW

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 1]
```diff
@@ -1 +1,941 @@
...
+    }
+
+    /// Get a reference to the node weight.
+    pub(crate) fn node(&self, node: NodeIndex) -> &FetchNode {
+        &self.graph[node]
+    }
+
+    /// Clone for saving a completed candidate; drops the undo log, which
+    /// snapshots never roll back.
+    pub(crate) fn snapshot(&self) -> Self {
```
Comment (dariuszkuc) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:470]: do we need derive clone? or should this be dropped?

guess the diff is around the `undo_log` but it looks like this is not used at all?

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 2]
```diff
@@ -1 +1,928 @@
...
+    },
+    RootHop {
+        root_type: CompositeTypeDefinitionPosition,
+        merge_at: Vec<FetchDataPathElement>,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum InputContribution {
+    /// @key fields the parent sends to enter the child; drives input
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:58]: This assumes the parent node contains all the data for the condition/selection set, which may not be the case for recursively planned conditions. My understanding is that's a current limitation of this incremental planner, which means there's a correctness/satisfiability regression for such cases (but it's also my understanding that we don't believe this occurs often enough to block an initial release, and that we'll address the limitation later).

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 3]
```diff
@@ -1 +1,928 @@
...
+
+/// Lightweight fetch graph for BULB search. Trial branches are applied,
+/// scored, and undone on one instance via an append-only mutation log:
+/// `checkpoint()` marks a log position, `rollback(cp)` reverses back to it.
+#[derive(Clone, Debug)]
+pub(crate) struct FetchGraph {
+    graph: StableDiGraph<FetchNode, FetchEdgeWeight>,
+    /// Root groups keyed by subgraph name. One search plans one root kind,
+    /// so the kind is not part of the key; root hops never register here.
+    root_groups: HashMap<Arc<str>, NodeIndex>,
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Given that `field_signatures()` (code from a previous PR for computing selection set builder collisions that was moved to a later PR) can report false positives, it seems that we could have extra root nodes per key too (though it depends on how such collisions are handled). If that's the case, you'll want to handle `root_groups` more closely to how `entity_groups` is handled.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 4]
```diff
@@ -1 +1,928 @@
...
+        while self.undo_log.len() > cp.0 {
+            match self.undo_log.pop().unwrap() {
+                FetchGraphOp::AddNode {
+                    node_index,
+                    root_key,
+                    entity_key,
+                } => {
+                    self.graph.remove_node(node_index);
+                    if let Some(key) = root_key {
+                        self.root_groups.remove(&key);
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: As noted above, if `root_groups` can contain extra nodes like `entity_groups`, you'll want to do a check on the node index similar to what's done with `entity_groups` below.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 5]
```diff
@@ -1 +1,928 @@
...
+                        .selection_builder
+                        .restore_head(prev_head);
+                }
+            }
+        }
+    }
+
+    /// Add a `FetchNode`, logging for rollback. `root_key` registers a
+    /// root group; entity nodes claim their `entity_groups` slot if free.
+    fn insert_node(&mut self, node: FetchNode, root_key: Option<Arc<str>>) -> NodeIndex {
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: I'm confused at why you need to pass in `root_key` here. Couldn't you also just get the root key by checking if `node.kind` is `FetchGroupKind::Root` and looking at `node.subgraph`? Are there cases where that's different?

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 6]
```diff
@@ -1 +1,928 @@
...
+    }
+
+    /// Whether a directed edge from `parent` to `child` exists.
+    pub(crate) fn has_edge(&self, parent: NodeIndex, child: NodeIndex) -> bool {
+        self.find_edge(parent, child).is_some()
+    }
+
+    /// Find the edge index for a directed edge from `parent` to `child`.
+    pub(crate) fn find_edge(&self, parent: NodeIndex, child: NodeIndex) -> Option<EdgeIndex> {
+        self.graph
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Would be useful to describe why you can't just call `self.graph.find_edge()` here.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 7]
```diff
@@ -1 +1,928 @@
...
+            .edges_directed(parent, Direction::Outgoing)
+            .find(|e| e.target() == child)
+            .map(|e| e.id())
+    }
+
+    /// Create a parent->child dependency edge with the given inputs. The
+    /// only way to create edges.
+    pub(crate) fn add_dependency(
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:378]: In addition to the caller needing to check for themselves that `!self.is_reachable(child, parent)`, do they also need to check that `!self.has_edge(parent, child)`? (To avoid adding a second edge between `parent` and `child`.) If so, would be good to have a `debug_assert!()` here for that.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 8]
```diff
@@ -1 +1,928 @@
...
+
+    /// Structural cost: FETCH_COST per group, scaled by pipeline depth
+    /// (longest parent chain). Recomputed per call; incremental caching is
+    /// deferred until the search is proven correct.
+    pub(crate) fn cost(&self) -> QueryPlanCost {
+        let Ok(order) = petgraph::algo::toposort(&self.graph, None) else {
+            debug_assert!(false, "cycle in fetch graph");
+            return f64::MAX;
+        };
+        let mut depth = vec![0u32; self.graph.node_bound()];
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:510]: My understanding is that `StableGraph` can reuse old node indices from deleted nodes (so `node_bound()` doesn't just always grow as you delete/add nodes), but that if `self.graph` balloons to some large size while exploring a branch and then shrinks back down, the `node_bound()` will still be high (and every call to `cost()` here would need to allocate that on the heap). Can this just be e.g. a `HashMap<NodeIndex, usize>` instead, or does the `Vec` make a significant performance difference in practice?

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 9]
```diff
@@ -1 +1,928 @@
...
+        let mut total: QueryPlanCost = 0.0;
+        for node in order {
+            let d = self
+                .graph
+                .edges_directed(node, Direction::Incoming)
+                .map(|e| depth[e.source().index()] + 1)
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: May want to use saturating add here to avoid wraparound on overflow (current type is `u32`).

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 1 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:470]: I think it makes sense to keep the derived clone. Skipping the undo log here is probably an unnecessary optimization. Simplified to `self.clone()` in 918dc3f.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs [hunk 10]
```diff
@@ -1 +1,928 @@
...
+#[derive(Clone, Debug)]
+pub(crate) struct FetchGraph {
+    graph: StableDiGraph<FetchNode, FetchEdgeWeight>,
+    /// Root groups keyed by subgraph name. One search plans one root kind,
+    /// so the kind is not part of the key; root hops never register here.
+    root_groups: HashMap<Arc<str>, NodeIndex>,
+    /// First-created entity group per (subgraph, merge_at), so group reuse
+    /// is a lookup instead of a node scan. The first node with a key owns
+    /// the slot; LIFO undo releases it with that node.
+    entity_groups: HashMap<EntityGroupKey, NodeIndex>,
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Seems like there's nothing for allowing root hop node reuse (though it's also something we can punt to later, given it doesn't affect plan correctness and assuming chances for reuse are low in practice).

**Review by sachindshinde — CHANGES_REQUESTED** [review]
As a general note, it would be nice if there were docstrings on each field in a struct for the nodes/edges.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 2 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:58]: This is just creating some of the infra for supporting required fields. Actual support for it comes in https://github.com/apollographql/router/pull/10139. For recursive conditions, we insert a dependency edge from the required field fetch to the dependent node, so those required fields will be available once the responses are flattened into the parent node.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 6 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: No reason. It was left over from an earlier iteration. Fixed in 02046b317 to delegate to petgraph's find_edge

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 7 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:378]: Added the `debug_assert!` in 02046b317. Callers (wire_key_edge, add_ordering_dependency in later PRs) check-then-reuse, so the invariant should hold.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 3 above)
Comment (dariuszkuc) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Still outstanding, and worth flagging that the asymmetry has grown rather than closed. At the chain tip (`915e1d784`) the `entity_groups` side picked up three defences and `root_groups` picked up none:

| | `entity_groups` | `root_groups` |
| --- | --- | --- |
| claim on insert | `!contains_key` guard — first claimant wins, never overwrites (`:334-337`) | unconditional `insert` (`:327`) |
| undo | dedicated `RegisterEntityGroup` op, pushed only when the slot was actually claimed (`:155`, `:302-304`) | unconditional `remove` inside the `AddNode` arm (`:271`), regardless of who owns the slot |
| read | liveness check + stale-slot eviction (`:479-484`) | bare `get` (`:361`) |

So the thing you pointed at as the model to follow has itself moved further ahead. The `AddNode` doc at `:127` also still encodes the unchecked assumption — *"a root node (`root_key` set) also removes **its** root_groups entry"* — where "its" is precisely what nothing verifies.

To be clear on severity: still latent. `get_or_create_root_group_with_defer` remains the only caller passing `Some(root_key)` and it checks first, so no current path can overwrite a live slot or remove someone else's. This is robustness, not a live bug.

The cheap fix is to mirror what the entity side already does — a `RegisterRootGroup { key }` op plus a `!contains_key` guard in `insert_node`. That also resolves your separate question in https://github.com/apollographql/router/pull/10102#discussion_r4007792384: with the key carried on its own op, `insert_node` no longer needs the `root_key` parameter at all, and the key is derivable from `node.kind` + `node.subgraph` + `node.defer_ref` anyway now that `FetchNode` carries `defer_ref`.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 8 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs:510]: I haven't benchmarked this specific version, but when I was implementing the [transitive reduction optimizations](https://github.com/apollographql/router/commit/05f8433f9d2f85a15b513356cd0af5c058552ec2#diff-0e5fb2879d5e5ee8ea259f971d3f2814a7a20258f33fca619ce9fc3d8dd536c3) a while back, I did see a large performance improvement to using `Vec` over `HashMap<NodeIndex>`. I think the cache locality trumps the unused allocations in these cases.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 9 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Done in 02046b317.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 3 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Reworked in 3a9c9a201: root, entity, and root hop groups now share a single keyed registry derived from the node itself, with one claim-if-free path, a dedicated `RegisterGroup` undo op logged only when the slot was actually claimed, and liveness-checked lookups. Since all three kinds ride the same mechanism, the root and entity sides can't drift apart again. There's also a regression test showing that a duplicate node's rollback leaves the owner's registration intact.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 4 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Covered by the same rework (3a9c9a201): lookups now go through `registered_group`, which debug_asserts the node is live. Registration is released by its own undo op in lock-step with node removal, so a live entry implies a live node; the assert documents that invariant rather than papering over stale entries.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 5 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Good call, removed in 3a9c9a201: `insert_node` now derives the key from the node's kind and subgraph, so there are no cases where a passed-in key could disagree with the node. One heads-up for later in the stack: connector-backed root nodes must not register (each connector resolution is deliberately its own fetch), so when connectors land the derivation grows a `connector.is_none()` guard instead of reintroducing the parameter.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs (hunk 10 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/mod.rs]: Went ahead and implemented it rather than punting: root hops now reuse a group per (subgraph, root kind, merge_at) via the same get-or-create + reachability-fallback shape entity groups use. The get-or-create lands here (3a9c9a201); the commit-logic PR wires it in and shares the cycle-check predicate with the entity path (563964ce4), so sibling fields resolved through one hop share a fetch, while a hop that transitively depends on the shared group (e.g. through a @requires chain back into the same subgraph) still gets a fresh node. Integration coverage: sibling sharing in the root-hop test on the integration-tests PR, and a requires-chain cycle guard test coming with the @requires PR.

**Review by sachindshinde — APPROVED** [review]

**Review by dariuszkuc — APPROVED** [review]
